### PR TITLE
Change `/labelValues` to match how the data is used

### PR DESCRIPTION
**Pull request by willr3**
This includes the changes mentioned in #1092 

The goal is to make the `/api/test/labelValues` endpoint easier for external tools (e.g. Grafana)
returning a map of `{ [labelName] : labelValue, ...}` is easier for grafana to process than `[ { name: labelName, value: labelValue},...]`
This also adds `runId` and `datasetId` to the object for each dataset. The id's are primarily for correlation.
We also add a `filter` to reduce the number of datasets in the result set.
Filters can be a json sub object `{[labelName]: labelValue,...}` where each `labelValue` must match for the dataset to be included.
Filters can also be a valid jsonpath expression where the dataset will be included if the expression returns a non-null value. There are valid jsonpath expressions that always return a boolean (t or f) which would result in all datasets (no filtering) we might want to add support for those type of expressions.

The `sort` option is currently limited to `labelNames`
`direction` is either Ascending or Descending
`limit` is the number of datasets to include (default is ALL)
`page` is the number of pages to skip.

This is still a work in progress but open for discussion. Some of the design considerations:
- [ ] Do we return an error if the filter is invalid? right now an invalid filter is ignored
- [ ] Do we support sorting by jsonpath or just key names
- [ ] Do we include the run `start` and `stop` along with the `runId` and `datasetId` for added correlation
- [ ] Do we add sort / filtering options for `start` and `stop`

### CODE REVIEW

**Comment by willr3** [comment]
Another rebase and squash, I think the endpoint changes are at a point where they can be merged.
I still think more changes are necessary but it would be good to put the new `{ [labelName]: labelValue,...}` into the test server to see how it improves integegrations.

**Comment by johnaohara** [comment]
@willr3 the CI is failing on this test: https://github.com/Hyperfoil/Horreum/actions/runs/7820024426/job/21333757495?pr=1282#step:6:3582

**Comment by willr3** [comment]
> @willr3 the CI is failing on this test: https://github.com/Hyperfoil/Horreum/actions/runs/7820024426/job/21333757495?pr=1282#step:6:3582

Should be all fixed now. I had to undo some changes where I used `TestDAO t = TestMapper.to(get(testId,null));` instead of `TestDAO t = TestDAO.findById(testId);` and I'm not sure why converting from `Test` to `TestDAO` breaks parts of Horreum.

File: horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/changes/Target.java [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import com.fasterxml.jackson.databind.node.ObjectNode;
 
 public class Target {
+   //concatenated simicolons varibleID;{fingeprintJson}
```
[the remark this answers is not in the document]
Comment (willr3) [reply horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/changes/Target.java:6]: I think we need some documentation on the *magic* strings until we get around to replacing them with typed or structured information. I could make it a `/**...*/` java code comment but I thought a quick note on "this is what I find goes into the variable" was good for improving code documentation

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 2]
```diff
@@ -61,16 +59,48 @@ public class TestServiceImpl implements TestService {
...
+   //TODO ASC or DESC based on input or two separate prepared statements?
    //@formatter:off
+   //TODO use ?& :filter to support finding top level keys?
+   protected static final String LABEL_VALUES_FILTER_CONTAINS_JSON = "where combined.values @> :filter";
+   //TODO use @? instead of jsonb_path_match
+   //@? tricks hibernate?
+   //Cannot mix parameter styles between JDBC-style, ordinal and named in the same query
+   //alternative is jsonb_path_exists
+   //a solution does exist! https://github.com/spring-projects/spring-data-jpa/issues/2551
+   //use \\?\\? to turn into a ? in the query... makes perfect sense
```
[the remark this answers is not in the document]
Comment (willr3) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:71]: I can clean up. The `//TODO` are about supporting alternative filters and should probably turn into discussions or issues

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 3]
```diff
@@ -61,16 +59,48 @@ public class TestServiceImpl implements TestService {
...
+/*         SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value) FILTER (WHERE label.name IS NOT NULL), '{}'::::jsonb) AS values, runId, dataset.id AS datasetId
          FROM dataset
          LEFT JOIN label_values lv ON dataset.id = lv.dataset_id
          LEFT JOIN label ON label.id = lv.label_id
          WHERE dataset.testid = ?1
             AND (label.id IS NULL OR (?2 AND label.filtering) OR (?3 AND label.metrics))
-         GROUP BY dataset_id
-         """;
+         GROUP BY dataset.id, runId
+         """;*/
```
[the remark this answers is not in the document]
Comment (willr3) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:94]: yes, I'll do a cleanup

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 4]
```diff
@@ -480,26 +510,159 @@ public void updateFolder(int id, String folder) {
...
+            //an attempt to see if the user was trying to provide json
+            if(filter!=null && filter.startsWith("{") || filter.endsWith("}")){
+               CheckResult jsonbResult = castCheck(filter,"jsonb",em);
+               if(!jsonbResult.ok()){
+                  //we expect this error, what do we do with it?
+               }else{
+                  //this would be a surprise and quite a problem :)
+               }
+            }else{
+               //what do we do about the invalid jsonpath?
```
[the remark this answers is not in the document]
Comment (willr3) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:631]: `castCheck` ensures the input string is a valid jsonpath syntax. It does not make sure it finds an entry in the document but that is what filtering is for :)

File: horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/data/ExportedLabelValues.java [hunk 5]
```diff
@@ -1,30 +1,54 @@
...
         List<ExportedLabelValues> fps = new ArrayList<>();
-        nodes.forEach( n -> fps.add( new ExportedLabelValues( FingerprintValue.parse(n))));
+        nodes.forEach(objects->{
+            JsonNode node = (JsonNode)objects[0];
+            Integer runId = Integer.parseInt(objects[1]==null?"-1":objects[1].toString());
+            Integer datasetId = Integer.parseInt(objects[2]==null?"-1":objects[2].toString());
+            if(node.isObject()){
+                fps.add(new ExportedLabelValues((ObjectNode) node,runId,datasetId));
+            }else{
+                //TODO alert that something is wrong in the db response
```
[the remark this answers is not in the document]
Comment (willr3) [reply horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/data/ExportedLabelValues.java:49]: Now I remember talking about the method return type (currently `List<ExportedLabelValues>`) not being able to express a return along with an error. I hope to fix that in a subsequent change.

**Comment by willr3** [comment]
I re-ran the tests 3 times (no code changes) because of intermittent failures in test cases. I think our existing tests are non-deterministic.

**Pull request merged**